// @gravity-ui/date-components/dist/esm/components/RangeDateField/utils/getRangeEditableSections.js
import { getEditableSections, isEditableSection, toEditableSection } from '../../DateField/utils';
export function getRangeEditableSections(sections, value, validSegments, delimiter) {
    const start = getEditableSections(sections, value.start, validSegments.start);
    const end = getEditableSections(sections, value.end, validSegments.end);
    const last = start[start.length - 1];
    let position = last.end;
    const previousEditableSection = last.nextEditableSection;
    const sectionsCount = start.length + 1;
    const delimiterSection = toEditableSection({
        type: 'literal',
        contentType: 'letter',
        format: delimiter,
        placeholder: delimiter,
        hasLeadingZeros: false,
    }, value.start, validSegments.start, position, previousEditableSection);
    position += delimiterSection.textValue.length - 1;
    let nextEditableSection;
    for (let index = 0; index < end.length; index++) {
        const section = end[index];
        section.start += position;
        section.end += position;
        if (section.previousEditableSection === 0 && nextEditableSection === undefined) {
            section.previousEditableSection = previousEditableSection;
        }
        else {
            section.previousEditableSection += sectionsCount;
        }
        section.nextEditableSection += sectionsCount;
        if (isEditableSection(section) && nextEditableSection === undefined) {
            nextEditableSection = index + sectionsCount;
        }
    }
    if (nextEditableSection !== undefined) {
        delimiterSection.nextEditableSection = nextEditableSection;
        start[previousEditableSection].nextEditableSection = nextEditableSection;
    }
    return [...start, delimiterSection, ...end];
}
