// @gravity-ui/date-components/dist/cjs/components/RangeDateField/utils/getRangeEditableSections.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRangeEditableSections = void 0;
const utils_1 = require("../../DateField/utils");
function getRangeEditableSections(sections, value, validSegments, delimiter) {
    const start = (0, utils_1.getEditableSections)(sections, value.start, validSegments.start);
    const end = (0, utils_1.getEditableSections)(sections, value.end, validSegments.end);
    const last = start[start.length - 1];
    let position = last.end;
    const previousEditableSection = last.nextEditableSection;
    const sectionsCount = start.length + 1;
    const delimiterSection = (0, utils_1.toEditableSection)({
        type: 'literal',
        contentType: 'letter',
        format: delimiter,
        placeholder: delimiter,
        hasLeadingZeros: false,
    }, value.start, validSegments.start, position, previousEditableSection);
    position += delimiterSection.textValue.length - 1;
    let nextEditableSection;
    for (let index = 0; index < end.length; index++) {
        const section = end[index];
        section.start += position;
        section.end += position;
        if (section.previousEditableSection === 0 && nextEditableSection === undefined) {
            section.previousEditableSection = previousEditableSection;
        }
        else {
            section.previousEditableSection += sectionsCount;
        }
        section.nextEditableSection += sectionsCount;
        if ((0, utils_1.isEditableSection)(section) && nextEditableSection === undefined) {
            nextEditableSection = index + sectionsCount;
        }
    }
    if (nextEditableSection !== undefined) {
        delimiterSection.nextEditableSection = nextEditableSection;
        start[previousEditableSection].nextEditableSection = nextEditableSection;
    }
    return [...start, delimiterSection, ...end];
}
exports.getRangeEditableSections = getRangeEditableSections;
