// @gravity-ui/date-components/dist/esm/components/RelativeRangeDatePicker/components/PickerDialog/PickerDialog.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { dateTime } from '@gravity-ui/date-utils';
import { Button, Popup, Sheet } from '@gravity-ui/uikit';
import { block } from '../../../../utils/cn';
import { RelativeDatePicker } from '../../../RelativeDatePicker';
import { Presets } from '../Presets/Presets';
import { Zones } from '../Zones/Zones';
import { i18n } from './i18n';
import { useRelativeRangeDatePickerDialogState } from './useRelativeRangeDatePickerDialogState';
import './PickerDialog.css';
const b = block('relative-range-date-picker-dialog');
export function PickerDialog({ props, state, open, onClose, isMobile, anchorRef, className, }) {
    if (isMobile) {
        return (_jsx(Sheet, { visible: open, onClose: onClose, contentClassName: b('content', { mobile: true, size: 'xl' }, className), children: _jsx(DialogContent, Object.assign({}, props, { size: "xl", state: state, onApply: onClose })) }));
    }
    return (_jsx(Popup, { open: open, onClose: onClose, role: "dialog", anchorRef: anchorRef, contentClassName: b('content', { size: props.size }, className), focusTrap: true, restoreFocus: true, children: _jsx(DialogContent, Object.assign({}, props, { state: state, onApply: onClose })) }));
}
function DialogContent(props) {
    var _a, _b, _c, _d, _e, _f, _g;
    const state = useRelativeRangeDatePickerDialogState(props.state, props);
    const placeholderValue = ((_a = props.placeholderValue) === null || _a === void 0 ? void 0 : _a.timeZone(props.state.timeZone)) ||
        dateTime({ timeZone: props.state.timeZone });
    const fieldProps = {
        timeZone: props.state.timeZone,
        format: props.format,
        minValue: props.minValue,
        maxValue: props.maxValue,
        hasClear: props.allowNullableValues,
        readOnly: props.readOnly,
        size: props.size,
        errorPlacement: 'inside',
    };
    return (_jsxs("div", { children: [_jsxs("div", { className: b('pickers'), children: [_jsx(RelativeDatePicker, Object.assign({}, fieldProps, { validationState: ((_b = state.startValidation) === null || _b === void 0 ? void 0 : _b.isInvalid) ? 'invalid' : undefined, errorMessage: ((_d = (_c = state.startValidation) === null || _c === void 0 ? void 0 : _c.errors) === null || _d === void 0 ? void 0 : _d.join('\n')) || i18n('Value is incorrect.'), placeholderValue: placeholderValue.startOf('day'), label: i18n('From'), value: state.start, onUpdate: state.setStart })), _jsx(RelativeDatePicker, Object.assign({}, fieldProps, { validationState: ((_e = state.endValidation) === null || _e === void 0 ? void 0 : _e.isInvalid) ? 'invalid' : undefined, errorMessage: ((_g = (_f = state.endValidation) === null || _f === void 0 ? void 0 : _f.errors) === null || _g === void 0 ? void 0 : _g.join('\n')) || i18n('Value is incorrect.'), placeholderValue: placeholderValue.endOf('day'), label: i18n('To'), value: state.end, onUpdate: state.setEnd, roundUp: true }))] }), props.withApplyButton && !props.readOnly ? (_jsx(Button, { disabled: state.isInvalid, size: props.size, onClick: () => {
                    state.applyValue();
                    props.onApply();
                }, className: b('apply'), width: "max", children: i18n('Apply') })) : null, props.withPresets && !props.readOnly ? (_jsx(Presets, { size: props.size, presetTabs: props.presetTabs, onChoosePreset: (start, end) => {
                    state.setRange({ type: 'relative', value: start }, { type: 'relative', value: end });
                }, minValue: props.minValue })) : null, props.withZonesList ? (_jsx("div", { className: b('zone'), children: _jsx(Zones, { value: state.timeZone, onUpdate: state.setTimeZone, disabled: props.readOnly, size: props.size }) })) : null] }));
}
