// @gravity-ui/date-components/dist/cjs/components/RelativeDateField/hooks/useRelativeDateFieldProps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRelativeDateFieldProps = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
function useRelativeDateFieldProps(state, props) {
    const [prevCorrectDate, setPrevCorrectDate] = react_1.default.useState(state.lastCorrectDate);
    const [focusedDate, setFocusedDate] = react_1.default.useState(state.lastCorrectDate);
    if (prevCorrectDate !== state.lastCorrectDate) {
        setPrevCorrectDate(state.lastCorrectDate);
        setFocusedDate(state.lastCorrectDate);
    }
    return {
        inputProps: {
            size: props.size,
            autoFocus: props.autoFocus,
            value: state.text,
            onUpdate: state.setText,
            disabled: state.disabled,
            hasClear: props.hasClear,
            validationState: state.validationState,
            errorMessage: props.errorMessage,
            errorPlacement: props.errorPlacement,
            label: props.label,
            id: props.id,
            startContent: props.startContent,
            endContent: props.endContent,
            pin: props.pin,
            view: props.view,
            placeholder: props.placeholder,
            onKeyDown: props.onKeyDown,
            onKeyUp: props.onKeyUp,
            onBlur: props.onBlur,
            onFocus: props.onFocus,
            controlProps: {
                'aria-label': props['aria-label'] || undefined,
                'aria-labelledby': props['aria-labelledby'] || undefined,
                'aria-describedby': props['aria-describedby'] || undefined,
                'aria-details': props['aria-details'] || undefined,
                'aria-disabled': state.disabled || undefined,
            },
        },
        calendarProps: {
            size: props.size === 's' ? 'm' : props.size,
            readOnly: true,
            value: state.parsedDate,
            focusedValue: focusedDate,
            onFocusUpdate: setFocusedDate,
        },
        timeInputProps: {
            size: props.size,
            readOnly: true,
            value: state.lastCorrectDate,
            format: 'LTS',
        },
    };
}
exports.useRelativeDateFieldProps = useRelativeDateFieldProps;
