// @gravity-ui/uikit/build/esm/components/useList/components/ListItemView/ListItemView.js
import { __rest } from "tslib";
import React from 'react';
import { Check, ChevronDown, ChevronUp } from '@gravity-ui/icons';
import { Icon } from '../../../Icon';
import { Text, colorText } from '../../../Text';
import { Flex, spacing } from '../../../layout';
import { block } from '../../../utils/cn';
import { LIST_ITEM_DATA_ATR, modToHeight } from '../../constants';
import './ListItemView.css';
const b = block('list-item-view');
export const ListItemViewSlot = (_a) => {
    var { children, indentation: indent = 1, className } = _a, props = __rest(_a, ["children", "indentation", "className"]);
    return (React.createElement(Flex, Object.assign({ width: indent * 16, className: b('slot', className) }, props), children));
};
const renderSafeIndentation = (indentation) => {
    if (indentation && indentation >= 1) {
        return (React.createElement(ListItemViewSlot, { indentation: Math.floor(indentation) }));
    }
    return null;
};
export const ListItemView = React.forwardRef(function ListItemView(_a, ref) {
    var { id, as: asProps, size = 'm', active, selected, disabled, activeOnHover: propsActiveOnHover, className, hasSelectionIcon = true, indentation, startSlot, subtitle, endSlot, title, height, expanded, dragging, style, role = 'option', onClick: _onClick } = _a, rest = __rest(_a, ["id", "as", "size", "active", "selected", "disabled", "activeOnHover", "className", "hasSelectionIcon", "indentation", "startSlot", "subtitle", "endSlot", "title", "height", "expanded", "dragging", "style", "role", "onClick"]);
    const as = asProps || 'div';
    const isGroup = typeof expanded === 'boolean';
    const onClick = disabled ? undefined : _onClick;
    const activeOnHover = typeof propsActiveOnHover === 'boolean' ? propsActiveOnHover : Boolean(onClick);
    return (React.createElement(Flex, Object.assign({ [LIST_ITEM_DATA_ATR]: id, role: role, "aria-selected": selected, onClick: onClick, className: b({
            active: dragging || active,
            selected: selected && !hasSelectionIcon,
            activeOnHover,
            radius: size,
            dragging,
            clickable: Boolean(onClick),
        }, spacing({ px: 2 }, className)), style: Object.assign({ minHeight: height !== null && height !== void 0 ? height : modToHeight[size][Number(Boolean(subtitle))] }, style), as: as, ref: ref, alignItems: "center", gap: "4", justifyContent: "space-between" }, rest),
        React.createElement(Flex, { gap: "2", alignItems: "center", grow: true },
            hasSelectionIcon && (React.createElement(ListItemViewSlot // reserve space
            , null, selected ? (React.createElement(Icon, { data: Check, size: 16, className: colorText({ color: 'info' }) })) : null)),
            renderSafeIndentation(indentation),
            isGroup ? (React.createElement(Icon, { className: b('icon', colorText({ color: disabled ? 'hint' : undefined })), data: expanded ? ChevronDown : ChevronUp, size: 16 })) : null,
            startSlot,
            React.createElement("div", { className: b('main-content') },
                typeof title === 'string' ? (React.createElement(Text, { ellipsis: true, color: disabled ? 'hint' : undefined, variant: isGroup ? 'subheader-1' : undefined }, title)) : (title),
                typeof subtitle === 'string' ? (React.createElement(Text, { ellipsis: true, color: disabled ? 'hint' : 'secondary' }, subtitle)) : (subtitle))),
        endSlot));
});
