// @gravity-ui/components/build/esm/components/CookieConsent/ConsentManager.js
import pick from 'lodash/pick';
import Cookies from 'universal-cookie';
export const COOKIE_NAME = 'analyticsConsents';
export const CONSENT_COOKIE_SETTINGS = {
    path: '/',
    maxAge: 60 * 60 * 24 * 365,
    secure: true,
    sameSite: true,
};
export var ConsentType;
(function (ConsentType) {
    ConsentType["Necessary"] = "necessary";
    ConsentType["Analytics"] = "analytics";
    ConsentType["Marketing"] = "marketing";
})(ConsentType || (ConsentType = {}));
export var ConsentMode;
(function (ConsentMode) {
    ConsentMode["Notification"] = "notification";
    ConsentMode["Manage"] = "manage";
    ConsentMode["Base"] = "base";
})(ConsentMode || (ConsentMode = {}));
export var AdditionalConsentParams;
(function (AdditionalConsentParams) {
    AdditionalConsentParams["Closed"] = "closed";
    AdditionalConsentParams["Edition"] = "edition";
})(AdditionalConsentParams || (AdditionalConsentParams = {}));
const cookies = new Cookies();
export class ConsentManager {
    constructor(mode, edition, cookieSettings = CONSENT_COOKIE_SETTINGS) {
        this.closed = false;
        this.consents = {};
        this.cookiesTypes = Object.values(ConsentType);
        this.subscribers = [];
        this.consentMode = mode;
        this.projectConsentEdition = edition;
        this.cookieSettings = cookieSettings;
        this.setInitValues();
    }
    get mode() {
        return this.consentMode;
    }
    get cookies() {
        return this.cookiesTypes;
    }
    get cookiesSettings() {
        return this.cookieSettings;
    }
    getConsents() {
        if (Object.keys(this.consents).length) {
            return this.consents;
        }
        return this.prepareConsent('OnlyNecessary');
    }
    subscribe(handler) {
        this.subscribers.push(handler);
        return () => {
            const index = this.subscribers.findIndex((value) => value === handler);
            if (index >= 0) {
                this.subscribers.splice(index, 1);
            }
        };
    }
    setConsents(values) {
        const consents = typeof values === 'string' ? this.prepareConsent(values) : values;
        const difference = Object.values(this.cookiesTypes).filter((type) => !consents[type] || consents[type] !== this.consents[type]);
        const differenceInVersion = this.consentEdition !== this.projectConsentEdition;
        const shouldClose = this.mode === ConsentMode.Notification && !this.closed;
        if (!difference.length && !differenceInVersion && !shouldClose) {
            return;
        }
        Object.assign(this.consents, consents);
        this.saveNewCookieValue();
        this.handleConsentChange(pick(consents, difference));
    }
    isConsentNotDefined() {
        if (this.mode === ConsentMode.Notification && !this.closed) {
            return true;
        }
        return !this.isAllConsentsDefined() || this.projectConsentEdition !== this.consentEdition;
    }
    prepareConsent(value) {
        return this.cookiesTypes.reduce((acc, type) => {
            acc[type] = value === 'All' ? true : type === ConsentType.Necessary;
            return acc;
        }, {});
    }
    isAllConsentsDefined() {
        return Object.values(this.cookiesTypes).every((type) => typeof this.consents[type] === 'boolean');
    }
    setInitValues() {
        const value = cookies.get(COOKIE_NAME);
        if (!(typeof value === 'object' && !Array.isArray(value) && value)) {
            return;
        }
        this.consents = Object.assign({}, pick(value, Object.values(ConsentType)));
        if (value[AdditionalConsentParams.Closed]) {
            this.closed = true;
        }
        if (value[AdditionalConsentParams.Edition]) {
            this.consentEdition = value.edition;
        }
    }
    saveNewCookieValue() {
        const newValue = Object.assign(Object.assign({}, this.consents), { [AdditionalConsentParams.Edition]: this.projectConsentEdition });
        this.consentEdition = this.projectConsentEdition;
        if (this.mode === ConsentMode.Notification) {
            newValue[AdditionalConsentParams.Closed] = true;
            this.closed = true;
            this.consents.closed = true;
        }
        cookies.set(COOKIE_NAME, newValue, this.cookieSettings);
    }
    handleConsentChange(changedConsents) {
        const allConsents = this.getConsents();
        this.subscribers.forEach((handler) => handler(changedConsents, allConsents));
    }
}
