// @gravity-ui/date-components/dist/cjs/components/RelativeRangeDatePicker/components/Presets/defaultPresets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allPresets = exports.DEFAULT_OTHERS_PRESETS = exports.DEFAULT_TIME_PRESETS = exports.DEFAULT_DATE_PRESETS = void 0;
const i18n_1 = require("./i18n");
exports.DEFAULT_DATE_PRESETS = [
    {
        from: 'now-1d',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last day');
        },
    },
    {
        from: 'now-3d',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 3 days');
        },
    },
    {
        from: 'now-1w',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last week');
        },
    },
    {
        from: 'now-1M',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last month');
        },
    },
    {
        from: 'now-3M',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 3 months');
        },
    },
    {
        from: 'now-6M',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 6 months');
        },
    },
    {
        from: 'now-1y',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last year');
        },
    },
    {
        from: 'now-3y',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 3 years');
        },
    },
];
exports.DEFAULT_TIME_PRESETS = [
    {
        from: 'now-5m',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 5 minutes');
        },
    },
    {
        from: 'now-15m',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 15 minutes');
        },
    },
    {
        from: 'now-30m',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 30 minutes');
        },
    },
    {
        from: 'now-1h',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last hour');
        },
    },
    {
        from: 'now-3h',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 3 hours');
        },
    },
    {
        from: 'now-6h',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 6 hours');
        },
    },
    {
        from: 'now-12h',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('Last 12 hours');
        },
    },
];
exports.DEFAULT_OTHERS_PRESETS = [
    {
        from: 'now/d',
        to: 'now/d',
        get title() {
            return (0, i18n_1.i18n)('Today');
        },
    },
    {
        from: 'now-1d/d',
        to: 'now-1d/d',
        get title() {
            return (0, i18n_1.i18n)('Yesterday');
        },
    },
    {
        from: 'now-2d/d',
        to: 'now-2d/d',
        get title() {
            return (0, i18n_1.i18n)('Day before yesterday');
        },
    },
    {
        from: 'now/w',
        to: 'now/w',
        get title() {
            return (0, i18n_1.i18n)('This week');
        },
    },
    {
        from: 'now/M',
        to: 'now/M',
        get title() {
            return (0, i18n_1.i18n)('This month');
        },
    },
    {
        from: 'now/y',
        to: 'now/y',
        get title() {
            return (0, i18n_1.i18n)('This year');
        },
    },
    {
        from: 'now/d',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('From start of day');
        },
    },
    {
        from: 'now/w',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('From start of week');
        },
    },
    {
        from: 'now/M',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('From start of month');
        },
    },
    {
        from: 'now/y',
        to: 'now',
        get title() {
            return (0, i18n_1.i18n)('From start of year');
        },
    },
];
exports.allPresets = exports.DEFAULT_TIME_PRESETS.concat(exports.DEFAULT_DATE_PRESETS, exports.DEFAULT_OTHERS_PRESETS);
