// @gravity-ui/components/build/cjs/components/PlaceholderContainer/PlaceholderContainer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaceholderContainer = void 0;
const react_1 = __importDefault(require("react"));
const cn_1 = require("../utils/cn");
const PlaceholderContainerAction_1 = require("./PlaceholderContainerAction");
const constants_1 = require("./constants");
require("./PlaceholderContainer.css");
const b = (0, cn_1.block)(constants_1.componentClassName);
class PlaceholderContainer extends react_1.default.Component {
    render() {
        const { direction, align, size } = this.props;
        const className = this.props.className || b();
        return (react_1.default.createElement("div", { className: b({ direction, align, size }, [className]) },
            react_1.default.createElement("div", { className: b('body') },
                react_1.default.createElement("div", { className: b('image', { size }) }, this.renderImage()),
                this.renderContent())));
    }
    renderImage() {
        if (typeof this.props.image === 'object' && 'url' in this.props.image) {
            const image = this.props.image;
            return react_1.default.createElement("img", { src: image.url, alt: image.alt || '' });
        }
        return this.props.image;
    }
    renderContent() {
        const { size } = this.props;
        const content = this.props.renderContent ? (this.props.renderContent()) : (react_1.default.createElement(react_1.default.Fragment, null,
            this.renderTitle(),
            this.renderDescription()));
        return (react_1.default.createElement("div", { className: b('content', { size }) },
            content,
            this.renderAction()));
    }
    renderTitle() {
        const { title } = this.props;
        if (!title) {
            return null;
        }
        return react_1.default.createElement("div", { className: b('title') }, title);
    }
    renderDescription() {
        const { description } = this.props;
        if (!description) {
            return null;
        }
        return react_1.default.createElement("div", { className: b('description') }, description);
    }
    renderAction() {
        const { action, renderAction } = this.props;
        if (renderAction) {
            return renderAction();
        }
        if (!action) {
            return null;
        }
        if (Array.isArray(action)) {
            if (!action.length) {
                return null;
            }
            return (react_1.default.createElement("div", { className: b('actions') }, action.map((actionItem) => (react_1.default.createElement(PlaceholderContainerAction_1.PlaceholderContainerAction, { key: actionItem.text, action: actionItem })))));
        }
        return (react_1.default.createElement("div", { className: b('actions') },
            react_1.default.createElement(PlaceholderContainerAction_1.PlaceholderContainerAction, { action: action })));
    }
}
exports.PlaceholderContainer = PlaceholderContainer;
PlaceholderContainer.Direction = constants_1.Direction;
PlaceholderContainer.Align = constants_1.Align;
PlaceholderContainer.defaultProps = {
    size: 'l',
    direction: constants_1.Direction.Row,
    align: constants_1.Align.Center,
};
