// @gravity-ui/date-components/dist/esm/components/RelativeRangeDatePicker/components/Presets/Presets.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { List, Tabs } from '@gravity-ui/uikit';
import { block } from '../../../../utils/cn';
import { PresetsDoc } from './PresetsDoc';
import { filterPresetTabs, getDefaultPresetTabs } from './utils';
import './Presets.css';
const b = block('relative-range-date-picker-presets');
export function Presets({ className, size = 'm', minValue, withTime, onChoosePreset, presetTabs, }) {
    var _a, _b;
    const tabs = React.useMemo(() => {
        return filterPresetTabs(presetTabs !== null && presetTabs !== void 0 ? presetTabs : getDefaultPresetTabs({ withTime }), { minValue });
    }, [withTime, minValue, presetTabs]);
    const [activeTabId, setActiveTab] = React.useState((_a = tabs[0]) === null || _a === void 0 ? void 0 : _a.id);
    if (tabs.length === 0) {
        return null;
    }
    const activeTab = (_b = tabs.find((tab) => tab.id === activeTabId)) !== null && _b !== void 0 ? _b : tabs[0];
    if (!activeTab) {
        return null;
    }
    if (activeTab.id !== activeTabId) {
        setActiveTab(activeTab.id);
    }
    return (_jsxs("div", { className: b({ size }, className), children: [_jsxs("div", { className: b('tabs'), children: [_jsx(Tabs, { activeTab: activeTabId, onSelectTab: setActiveTab, items: tabs, size: size === 's' ? 'm' : size }), _jsx(PresetsDoc, { className: b('doc'), size: size })] }), _jsx("div", { className: b('content'), children: _jsx(PresetsList, { presets: activeTab.presets, onChoosePreset: onChoosePreset, size: size }) })] }));
}
export const SIZE_TO_ITEM_HEIGHT = {
    s: 28,
    m: 28,
    l: 32,
    xl: 36,
};
function PresetsList({ presets, onChoosePreset, size = 'm' }) {
    const ref = React.useRef(null);
    React.useEffect(() => {
        var _a, _b;
        const list = ref.current;
        const container = (_b = (_a = ref.current) === null || _a === void 0 ? void 0 : _a.refContainer.current) === null || _b === void 0 ? void 0 : _b.node;
        if (list && container) {
            try {
                container.setAttribute('tabindex', '0');
                container.setAttribute('class', b('list-container'));
                const handleFocus = () => {
                    if (list.getActiveItem() === null) {
                        list.activateItem(0, true);
                    }
                };
                container.addEventListener('focus', handleFocus);
                return () => {
                    container.removeEventListener('focus', handleFocus);
                };
            }
            catch (_c) {
                // Oooops
            }
        }
        return undefined;
    }, []);
    return (_jsx(List, { ref: ref, className: b('list'), itemClassName: b('item'), items: presets, filterable: false, virtualized: false, renderItem: (item) => item.title, itemHeight: SIZE_TO_ITEM_HEIGHT[size], onItemClick: (item) => {
            onChoosePreset(item.from, item.to);
        } }));
}
